### PR TITLE
feat: add wildcard and deselection support to `--components`

### LINKED ISSUE #1794: A wildcard and exclude syntax for `--components`
### Is your feature request related to a problem? Please describe.
As a more advanced deployer I want the ability to more specifically control the components in a package that get deployed so that I can more quickly run a deployment from the CLI.

### Describe the solution you'd like

- **Given** I have a package with optional components
- **When** I run `zarf package deploy --components test-*`
- **Then** all of the components prefixed with `test-` will be deployed

--

- **Given** I have a package with optional components
- **And** some of them are `default: true`
- **When** I run `zarf package deploy --components test-`
- **Then** the default component named `test` will be excluded

### Additional context
In addition to partial wildcard, it would be nice to have full wildcard support and the ability to mix wildcards and exclusions.

### LINKED ISSUE #2035: Refactor `--components` to respect the full list in interactive mode
### Describe what should be investigated or refactored
Currently if you specify a list of optional components without `--confirm` you will be re-prompted for any components that you did not specify - this can lead to confusion and potential for errors to be made - Zarf should respect the full list of components if it is provided.

### Links to any relevant code
https://github.com/defenseunicorns/zarf/blob/main/src/pkg/packager/components.go#L20

### Additional context
This was brought up by a few users of Zarf as strange behavior to them and even though it would change things it would only be for interactive users so shouldn't be breaking.

### LINKED ISSUE #2051: Add an option to say "yes to all" when deploying a zarf package that has unrequired components
### Is your feature request related to a problem? Please describe.
If all of the components in my package are required, I can't use `zarf package deploy --components x,y --confirm` without it still deploying everything

If all of the components in my package are not required, I can't deploy everything without spamming yes through the prompts (this doesn't sound like a big deal, but I'm working with a very large package and it's easy to miss a step)

### Describe the solution you'd like
I'd like to be able to satisfy both conditions above with a single zarf package such that for day 1, I can deploy everything easily without spamming yes to prompts, but for day 2 I can use the components flag and deploy only the components I specify.

In this scenario, you'd specify that all components in your package are not-required, and _can still be prompted_, but you can also bypass the prompts and say yes to all

Something like:
```sh
zarf package deploy --yes-to-all
```

### Describe alternatives you've considered
Currently, we are using yaml anchors for the required field on the components and just editing the zarf.yaml to toggle required true/false to facilitate deploying everything initially and then using the components flag for updates

### PR DESCRIPTION
**Pull request by Racer159**
## Description

This adds wildcard and `default` exclusion support to the `--components` field

## Related Issue

Fixes #1794
Fixes #2051 
Fixes #2035

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Other (security config, docs update, etc)

## Checklist before merging

- [ ] Test, docs, adr added or updated as needed
- [x] [Contributor Guide Steps](https://github.com/defenseunicorns/zarf/blob/main/CONTRIBUTING.md#developer-workflow) followed

### CODE REVIEW

**Comment by Racer159** [comment]
`default` is not implemented yet and the docs need updating

**Comment by Racer159** [comment]
`mirror` and `remove` need to be normalized to use the same logic

**Comment by Racer159** [comment]
#2035 should be addressed by this issue as well

File: src/pkg/packager/components.go [hunk 1]
```diff
@@ -110,40 +156,43 @@ func (p *Packager) validateRequests(validComponentsList []types.ZarfComponent, r
...
 
-	// All other cases, return false
-	return false
+	// All other cases we don't know if we should include or exclude yet
+	return false, false
 }
 
 // Confirm optional component.
-func (p *Packager) confirmOptionalComponent(component types.ZarfComponent) (confirmComponent bool) {
+func confirmOptionalComponent(component types.ZarfComponent) (confirmComponent bool) {
```
Comment (Noxsios) [inline src/pkg/packager/components.go:195]: can be moved to `interactive` package if not needed in `packager` anymore

File: src/pkg/packager/components.go [hunk 2]
```diff
@@ -169,7 +218,7 @@ func (p *Packager) confirmOptionalComponent(component types.ZarfComponent) (conf
 	return confirmComponent
 }
 
-func (p *Packager) confirmChoiceGroup(componentGroup []types.ZarfComponent) types.ZarfComponent {
+func confirmChoiceGroup(componentGroup []types.ZarfComponent) types.ZarfComponent {
```
Comment (Noxsios) [inline src/pkg/packager/components.go:221]: can be moved to `interactive` package if not needed in `packager` anymore

File: src/pkg/packager/components.go [hunk 3]
```diff
@@ -80,27 +93,60 @@ func (p *Packager) getValidComponents() []types.ZarfComponent {
...
 
+func (p *Packager) forRequestedComponents(function func(types.ZarfComponent) error) error {
+	requestedComponents := helpers.StringToSlice(p.cfg.PkgOpts.OptionalComponents)
+	partialMirror := len(requestedComponents) > 0 && requestedComponents[0] != ""
+
+	for _, component := range p.cfg.Pkg.Components {
+		included, excluded := false, false
+
+		if partialMirror {
+			included, excluded = includedOrExcluded(component, requestedComponents)
```
Comment (Noxsios) [inline src/pkg/packager/components.go:117]: if this function is meant to be used more generically, the `partialMirror` var is a confusing name. Perhaps `isPartial`?

**Review by Noxsios — CHANGES_REQUESTED** [review]

File: docs/3-create-a-zarf-package/2-zarf-components.md [hunk 4]
```diff
@@ -213,3 +213,21 @@ $ zarf package deploy ./path/to/package.tar.zst --confirm
...
+:::tip
+
+You can deploy components in a package using globbing as well.  The following would deploy all components regardless of optional status:
+
+```bash
+# deploy optional-component-1 and optional-component-2 components whether they are required or not
+$ zarf package deploy ./path/to/package.tar.zst --components=*
+```
+
+If you have any `default` components in a package definition you can also exclude those from the CLI with a trailing `-` (similar to removing a label from a resource with `kubectl`)
```
[the remark this answers is not in the document]
Comment (Racer159) [reply docs/3-create-a-zarf-package/2-zarf-components.md:226]: I based it on how kubectl removes labels - https://www.golinuxcloud.com/kubectl-label-node/

We could do ! but that is a special character in the shell

[the remark this answers is not in the document]
Comment (Racer159) [reply docs/3-create-a-zarf-package/2-zarf-components.md:217]: fine... you won

File: src/internal/packager/helm/post-render.go [hunk 5]
```diff
@@ -55,11 +55,10 @@ func (h *Helm) newRenderer() (*renderer, error) {
 
 func (r *renderer) Run(renderedManifests *bytes.Buffer) (*bytes.Buffer, error) {
 	// This is very low cost and consistent for how we replace elsewhere, also good for debugging
-	tempDir, err := utils.MakeTempDir(config.CommonOptions.TempDirectory)
+	tempDir, err := utils.MakeTempDir(r.chartPath)
 	if err != nil {
 		return nil, fmt.Errorf("unable to create tmpdir:  %w", err)
 	}
-	defer os.RemoveAll(tempDir)
```
[the remark this answers is not in the document]
Comment (Racer159) [reply src/internal/packager/helm/post-render.go:62]: yes - r.chartPath is inside the temp dir and will be cleaned if Zarf successfully runs - came from this convo: https://kubernetes.slack.com/archives/C03B6BJAUJ3/p1702042791854189

File: docs/3-create-a-zarf-package/2-zarf-components.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (Noxsios) [reply docs/3-create-a-zarf-package/2-zarf-components.md]: at what cost tho

**Review by Noxsios — CHANGES_REQUESTED** [review]

**Review by Noxsios — APPROVED** [review]

**Pull request merged**